libs/compact-map/src/Data/Compact/KeyMap.hs

```diff
@@ -248,67 +297,86 @@ insertWithKey' n0 combine path k x kmap = go n0 kmap
                   else Two bmap x0 st'
       where
         i = indexFromSegment bmap j
-        j = path !! n
-    go n t@(Full arr) =
+        j = path VP.! n
+    go !n t@(Full arr) =
       let !st = index arr i
           !st' = go (n + 1) st
        in if st' `ptrEq` st
             then t
             else Full (update arr i st')
       where
         i = indexFromSegment fullNodeMask j
-        j = path !! n
+        j = path VP.! n
+{-# INLINE insertWithKey' #-}
 
 twoLeaf :: Path -> KeyMap v -> Path -> Key -> v -> KeyMap v
-twoLeaf [] _ _ _ _ = error ("the path ran out of segments in twoLeaf case 1.")
-twoLeaf _ _ [] _ _ = error ("the path ran out of segments in twoLeaf case 1.")
-twoLeaf (i : is) leaf1 (j : js) k2 v2
-  | i == j = One i (twoLeaf is leaf1 js k2 v2)
-  | otherwise =
-    if i < j
-      then Two (setBits [i, j]) leaf1 (Leaf k2 v2)
-      else Two (setBits [i, j]) (Leaf k2 v2) leaf1
+twoLeaf path1 leaf1 path2 k2 v2 = go path1 path2
+  where
+    leaf2 = Leaf k2 v2
+    go p1 p2
+      | Just (i, is) <- VP.uncons p1,
+        Just (j, js) <- VP.uncons p2 =
+        if i == j
+          then One i (go is js)
+          else
+            let two = Two (setBits [i, j])
+             in if i < j
+                  then two leaf1 leaf2
+                  else two leaf2 leaf1
+      | otherwise =
+        error $
+          concat
+            [ "The path ran out of segments in 'twoLeaf'. \npath1: ",
+              show path1,
+              "\npath2: ",
+              show path2
+            ]
+{-# INLINE twoLeaf #-}
 
 insertWithKey :: (Key -> v -> v -> v) -> Key -> v -> KeyMap v -> KeyMap v
-insertWithKey f k v m = insertWithKey' 0 f (keyPath k) k v m
+insertWithKey f k = insertWithKey' 0 f (keyPath k) k
+{-# INLINE insertWithKey #-}
 
 insertWith :: (t -> t -> t) -> Key -> t -> KeyMap t -> KeyMap t
-insertWith f k v m = insertWithKey' 0 (\_ key val -> f key val) (keyPath k) k v m
+insertWith f k = insertWithKey' 0 (\_ key val -> f key val) (keyPath k) k
+{-# INLINE insertWith #-}
 
 insert :: Key -> v -> KeyMap v -> KeyMap v
-insert k v m = insertWithKey' 0 (\_key new _old -> new) (keyPath k) k v m
+insert !k = insertWithKey' 0 (\_key new _old -> new) (keyPath k) k
+{-# INLINE insert #-}
 
 fromList :: [(Key, v)] -> KeyMap v
-fromList ps = foldl' accum Empty ps
+fromList = foldl' accum Empty
   where
-    accum ans (k, v) = insert k v ans
+    accum !ans (k, v) = insert k v ans
+{-# INLINE fromList #-}
 
```

Comment:
It isn't technically necessary to fix memory leaks, but there is a subtle difference between just `foldl'` and `foldl'` with bang on accumulator: https://www.well-typed.com/blog/2014/04/fixing-foldl/#a-postscript-which-foldl

With all honesty I added it because I was benchmarking the function and I was tapping into all possible optimizations I could find. So this change is not strictly necessary (pun intended), but it won't hurt either.